tools/yulPhaser/PairSelections.cpp

```diff
@@ -0,0 +1,65 @@
+/*
+	This file is part of solidity.
+
+	solidity is free software: you can redistribute it and/or modify
+	it under the terms of the GNU General Public License as published by
+	the Free Software Foundation, either version 3 of the License, or
+	(at your option) any later version.
+
+	solidity is distributed in the hope that it will be useful,
+	but WITHOUT ANY WARRANTY; without even the implied warranty of
+	MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+	GNU General Public License for more details.
+
+	You should have received a copy of the GNU General Public License
+	along with solidity.  If not, see <http://www.gnu.org/licenses/>.
+*/
+
+#include <tools/yulPhaser/PairSelections.h>
+
+#include <tools/yulPhaser/SimulationRNG.h>
+
+#include <cmath>
+
+using namespace std;
+using namespace solidity::phaser;
+
+vector<tuple<size_t, size_t>> RandomPairSelection::materialise(size_t _poolSize) const
+{
+	if (_poolSize < 2)
+		return {};
+
+	size_t count = static_cast<size_t>(round(_poolSize * m_selectionSize));
+
+	vector<tuple<size_t, size_t>> selection;
+	for (size_t i = 0; i < count; ++i)
+	{
+		size_t index1 = SimulationRNG::uniformInt(0, _poolSize - 1);
+		size_t index2;
+		do
+		{
+			index2 = SimulationRNG::uniformInt(0, _poolSize - 1);
+		} while (index1 == index2);
```

Comment:
Is it actually important that they are different?

Reply:
I think so. Allowing them to be the same results in a higher chance that we'll get duplicates in the population. And the more duplicates there are the higher the chance that we'll get even more of them as a result of crossover.

Even with this extra loop the duplicates are still a problem though. Eventually I made the algorithm just replace duplicates with completely random chromosomes (and `--no-randomise-duplicates` option to turn it off) as suggested in the initial paper. But that's in #8423. Also, after reading up more on the theory I'm not fully convinced that this randomisation is a good thing - in the later stages of many genetic algorithms the population is supposed to get less and less diverse so that refinements get finer. Randomisation might interfere with that. I'll have to test that to be sure though.

If anything, I'd rather make it stricter - so that only pairs `i, j` (where `j < j`) can be selected.

There are algorithms where the duplicates don't matter because the whole population is replaced in each round. They may even be a feature, making the best chromosomes more likely to be selected for crossover. But if the elite is large (like in GEWEP - half the population) they just take up space and for these algorithms I think it's better if we keep them to a minimum.